lib/ecto/query/builder/select.ex

```diff
@@ -98,20 +99,24 @@ defmodule Ecto.Query.Builder.Select do
   end
 
   # map/struct(var, [:foo, :bar])
-  defp escape({tag, _, [{var, _, context}, fields]}, {params, take}, vars, env)
+  defp escape({tag, _, [{var, _, context}, fields]}, {params, take, subqueries}, vars, env)
        when tag in [:map, :struct] and is_atom(var) and is_atom(context) do
     taken = escape_fields(fields, tag, env)
     expr = Builder.escape_var!(var, vars)
     take = add_take(take, Builder.find_var!(var, vars), {tag, taken})
-    {expr, {params, take}}
+    {expr, {params, take, subqueries}}
   end
 
-  defp escape(expr, params_take, vars, env) do
-    Builder.escape(expr, :any, params_take, vars, {env, &escape_expansion/5})
+  defp escape(expr, {params, take, subqueries}, vars, env) do
+    {expr, {params, subqueries}} =
+      Builder.escape(expr, :any, {params, subqueries}, vars, {env, &escape_expansion/5})
```

Comment:
There is a bug hiding here and apparently we have no tests, but your change makes it so we ignore any "take" operation that happens inside expansion.

The correct approach should be to have `{params, take_subqueries}`, which are in the form of `{[], {%{}, []}}` instead of a triplet.

Reply:
I was already looking at this earlier. It _seems_ that `Builder.escape` never modifies the `acc` part of `{params, acc}`, except for adding subqueries. Any pointer as to how to produce the bug?

Reply:
It never modifies it but it passes it to expand_expansion, which may modify it. You can reproduce it by having a macro `my_macro(p)` that expands to `map(p, [:some, :field])`. The fields inside the expansion won’t be collected.

Reply:
I managed to produce the bug, add a test, and fix it in [4119e46](https://github.com/elixir-ecto/ecto/pull/3928/commits/4119e4602a2b036f053f39dac4329c6e8cbf2676). Can't say I fully understand it quite yet, which probably shows in the way I implemented it. What could I improve?